File: mock-agent.js
#!/usr/bin/env node

const fs = require('fs').promises;
const fsSync = require('fs');
const path = require('path');
const readline = require('readline');

// Agent personality traits
const PERSONALITIES = {
  Claude: {
    style: 'analytical and thorough',
    greeting: "Hello! I'm Claude. I'll analyze this challenge systematically.",
    focus: 'architecture and best practices'
  },
  Codex: {
    style: 'practical and code-focused',
    greeting: "Hey, Codex here. Let me think about the implementation details.",
    focus: 'concrete implementation'
  },
  Gemini: {
    style: 'creative and explorative',
    greeting: "Hi! Gemini joining. I'll explore alternative approaches.",
    focus: 'innovative solutions'
  }
};

class MockAgent {
  constructor() {
    this.agentName = process.env.AGENT_NAME || 'Agent';
    this.personality = PERSONALITIES[this.agentName] || PERSONALITIES.Claude;
    this.chatFile = null;
    this.planFile = null;
    this.workspaceDir = process.cwd();
    this.messageCount = 0;
    this.hasGreeted = false;
    this.discussionRounds = 0;
  }

  log(message) {
    console.error(`[${this.agentName}] ${message}`);
  }

  async initialize() {
    this.log('Initializing...');

    // Read the initial prompt from stdin
    const prompt = await this.readPrompt();
    this.log('Received prompt');

    // Parse workspace info from prompt
    const workspaceMatch = prompt.match(/workspace directory: (.+)/);
    if (workspaceMatch) {
      this.workspaceDir = workspaceMatch[1].trim();
    }

    const chatMatch = prompt.match(/shared chat file: (.+)/);
    if (chatMatch) {
      this.chatFile = path.join(this.workspaceDir, chatMatch[1].trim());
    } else {
      this.chatFile = path.join(this.workspaceDir, 'CHAT.md');
    }

    const planMatch = prompt.match(/final plan file: (.+)/);
    if (planMatch) {
      this.planFile = path.join(this.workspaceDir, planMatch[1].trim());
    } else {
      this.planFile = path.join(this.workspaceDir, 'PLAN_FINAL.md');
    }

    this.log(`Chat file: ${this.chatFile}`);
    this.log(`Plan file: ${this.planFile}`);

    // Extract the actual challenge
    const challengeMatch = prompt.match(/CHALLENGE:\s*(.+?)(?=\n\n|$)/s);
    if (challengeMatch) {
      this.challenge = challengeMatch[1].trim();
      this.log(`Challenge: ${this.challenge.substring(0, 50)}...`);
    }
  }

  async readPrompt() {
    return new Promise((resolve) => {
      let promptData = '';
      const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
        terminal: false
      });

      rl.on('line', (line) => {
        promptData += line + '\n';
      });

      // Set a timeout to finish reading
      setTimeout(() => {
        rl.close();
        resolve(promptData);
      }, 1000);
    });
  }

  async readChat() {
    try {
      return await fs.readFile(this.chatFile, 'utf8');
    } catch (error) {
      return '';
    }
  }

  async appendToChat(message) {
    const timestamp = new Date().toLocaleTimeString();
    const formattedMessage = `\n\n[${this.agentName} @ ${timestamp}]: ${message}\n`;

    try {
      await fs.appendFile(this.chatFile, formattedMessage);
      this.log('Appended message to chat');
    } catch (error) {
      this.log(`Error appending to chat: ${error.message}`);
    }
  }

  analyzeChat(chatContent) {
    const messages = chatContent.split('\n\n').filter(line =>
      line.includes('[') && line.includes(']:')
    );

    const myMessages = messages.filter(msg => msg.includes(`[${this.agentName}`)).length;
    const otherMessages = messages.length - myMessages;
    const hasUserMessage = chatContent.includes('[User @');

    return {
      totalMessages: messages.length,
      myMessages,
      otherMessages,
      hasUserMessage,
      shouldRespond: otherMessages > myMessages || hasUserMessage
    };
  }

  generateResponse(chatContent, analysis) {
    const responses = {
      greeting: [
        `${this.personality.greeting} Looking at this challenge about: ${this.challenge?.substring(0, 100)}...`,
        `I think we should focus on ${this.personality.focus}.`,
        `What do you all think?`
      ].join(' '),

      analysis: [
        `Based on what's been discussed, I have some thoughts on ${this.personality.focus}.`,
        this.getAnalysisInsight(),
        `Does this align with everyone's thinking?`
      ].join(' '),

      agreement: [
        `I agree with the points raised.`,
        `From my perspective focusing on ${this.personality.focus}, I'd add that ${this.getImplementationDetail()}.`
      ].join(' '),

      challenge: [
        `Interesting points, but I wonder if we've considered ${this.getChallengePoint()}.`,
        `This could be important for ${this.personality.focus}.`
      ].join(' '),

      summary: [
        `Let me summarize what we've discussed:`,
        this.getSummary(chatContent),
        `I think we have enough to create a final plan. Shall I draft it?`
      ].join(' '),

      plan: [
        `I'll create the final plan now based on our discussion.`
      ].join(' ')
    };

    // Determine which response to use
    if (!this.hasGreeted) {
      this.hasGreeted = true;
      return responses.greeting;
    }

    if (this.discussionRounds < 2) {
      this.discussionRounds++;
      return responses.analysis;
    }

    if (this.discussionRounds < 4) {
      this.discussionRounds++;
      return Math.random() > 0.5 ? responses.agreement : responses.challenge;
    }

    if (this.discussionRounds < 5) {
      this.discussionRounds++;
      return responses.summary;
    }

    return responses.plan;
  }

  getAnalysisInsight() {
    const insights = {
      Claude: 'we should consider scalability and maintainability from the start',
      Codex: 'the implementation should be straightforward and testable',
      Gemini: 'there might be some innovative patterns we could explore here'
    };
    return insights[this.agentName] || 'this needs careful consideration';
  }

  getImplementationDetail() {
    const details = {
      Claude: 'proper error handling and edge cases are critical',
      Codex: 'we should use well-tested libraries where possible',
      Gemini: 'we could experiment with a novel approach here'
    };
    return details[this.agentName] || 'attention to detail matters';
  }

  getChallengePoint() {
    const points = {
      Claude: 'the long-term maintenance implications',
      Codex: 'performance under load',
      Gemini: 'alternative architectures that might be more elegant'
    };
    return points[this.agentName] || 'potential issues down the road';
  }

  getSummary(chatContent) {
    return `We've discussed the challenge and covered ${this.personality.focus}. The consensus seems to be forming around a solid approach.`;
  }

  async writePlan() {
    const plan = `# Final Implementation Plan

## Challenge
${this.challenge || 'Multi-agent collaboration challenge'}

## Solution Overview
Based on our collaborative discussion, here's our recommended approach:

### 1. Architecture
- **Design Pattern**: Modular, scalable architecture
- **Key Components**:
  - Core logic layer
  - Data access layer
  - API/Interface layer
- **Rationale**: Ensures maintainability and testability

### 2. Implementation Strategy
- **Phase 1**: Set up core infrastructure and data models
- **Phase 2**: Implement business logic with proper error handling
- **Phase 3**: Build API layer with comprehensive validation
- **Phase 4**: Add monitoring, logging, and optimization

### 3. Technical Considerations
- **Performance**: Implement caching and optimize critical paths
- **Scalability**: Design for horizontal scaling from the start
- **Security**: Follow best practices for data validation and access control
- **Testing**: Comprehensive unit, integration, and end-to-end tests

### 4. Key Decisions
- Use proven, well-maintained libraries
- Prioritize code clarity and maintainability
- Implement proper logging and monitoring
- Plan for graceful degradation and error recovery

### 5. Next Steps
1. Set up development environment
2. Create project structure
3. Implement core functionality
4. Add tests and documentation
5. Deploy and monitor

## Conclusion
This plan represents the collective thinking of all agents, balancing ${PERSONALITIES.Claude.focus}, ${PERSONALITIES.Codex.focus}, and ${PERSONALITIES.Gemini.focus}.

---
*Generated by ${this.agentName} based on multi-agent collaboration*
`;

    try {
      await fs.writeFile(this.planFile, plan);
      this.log('Final plan written successfully');
    } catch (error) {
      this.log(`Error writing plan: ${error.message}`);
    }
  }

  async run() {
    await this.initialize();

    this.log('Starting main loop...');

    // Main loop
    while (true) {
      await this.sleep(3000); // Wait 3 seconds between checks

      try {
        const chatContent = await this.readChat();
        const analysis = this.analyzeChat(chatContent);

        this.log(`Analysis: ${analysis.totalMessages} messages, ${analysis.myMessages} mine`);

        // Decide whether to respond
        if (this.discussionRounds >= 5 && this.agentName === 'Claude') {
          // Claude writes the final plan
          await this.appendToChat('I\'ll write the final plan now based on our discussion.');
          await this.sleep(2000);
          await this.writePlan();
          this.log('Plan written, exiting...');
          break;
        } else if (analysis.shouldRespond && this.discussionRounds < 6) {
          // Generate and send response
          await this.sleep(1000 + Math.random() * 2000); // Random delay for realism
          const response = this.generateResponse(chatContent, analysis);
          await this.appendToChat(response);
        }
      } catch (error) {
        this.log(`Error in main loop: ${error.message}`);
      }
    }

    this.log('Agent finished');
  }

  sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

// Run the agent
const agent = new MockAgent();
agent.run().catch(error => {
  console.error('Fatal error:', error);
  process.exit(1);
});

File: mock_agent.py
#!/usr/bin/env python3
"""
Mock AI Agent for Testing the Multi-Agent Orchestrator

This simulates an AI agent that:
1. Reads the initial prompt from stdin
2. Periodically reads CHAT.md
3. Responds with simulated thinking and messages
4. Eventually contributes to PLAN_FINAL.md

Usage:
    python mock_agent.py <agent_name>

Example:
    python mock_agent.py Claude
"""

import sys
import time
import random
from pathlib import Path
from datetime import datetime


class MockAgent:
    def __init__(self, name: str):
        self.name = name
        self.chat_file = Path("CHAT.md")
        self.plan_file = Path("PLAN_FINAL.md")
        self.last_chat_content = ""
        self.message_count = 0
        self.has_proposed_plan = False

        # Different personalities for variety
        self.personalities = {
            "Claude": {
                "style": "analytical and thorough",
                "responses": [
                    "I think we should consider the architectural implications...",
                    "Let me break this down into components:",
                    "What about edge cases like...",
                    "I agree, but we should also validate...",
                ],
            },
            "Codex": {
                "style": "practical and code-focused",
                "responses": [
                    "Here's how I'd implement that:",
                    "We could use a pattern like...",
                    "From a performance standpoint...",
                    "Let me prototype this approach:",
                ],
            },
            "Gemini": {
                "style": "creative and questioning",
                "responses": [
                    "What if we approached it from a different angle?",
                    "I'm seeing some alternatives here...",
                    "Could we simplify this by...",
                    "Let's challenge that assumption:",
                ],
            },
        }

        self.personality = self.personalities.get(
            name, self.personalities["Claude"]
        )

    def log(self, message: str):
        """Log to stdout (visible in agent's tab)"""
        timestamp = datetime.now().strftime("%H:%M:%S")
        print(f"[{timestamp}] {message}", flush=True)

    def read_prompt(self):
        """Read initial prompt from stdin"""
        self.log(f"{self.name} agent starting...")
        self.log("Waiting for challenge prompt...")

        # Read all input (the initial prompt)
        prompt = sys.stdin.read()

        self.log(f"Received prompt ({len(prompt)} chars)")
        self.log(f"Challenge preview: {prompt[:100]}...")
        return prompt

    def read_chat(self) -> str:
        """Read current CHAT.md content"""
        if not self.chat_file.exists():
            return ""
        return self.chat_file.read_text()

    def append_to_chat(self, message: str):
        """Append a message to CHAT.md"""
        with open(self.chat_file, "a") as f:
            f.write(f"\n[{self.name}]: {message}\n")
        self.log(f"Posted to CHAT.md: {message[:50]}...")

    def analyze_chat(self, content: str) -> dict:
        """Analyze the chat content"""
        lines = content.strip().split("\n")
        messages = [
            line for line in lines if line.startswith("[") and "]: " in line
        ]

        return {
            "total_messages": len(messages),
            "my_messages": len([m for m in messages if m.startswith(f"[{self.name}]")]),
            "other_messages": len([m for m in messages if not m.startswith(f"[{self.name}]")]),
            "has_user_messages": any("[User" in m for m in messages),
        }

    def generate_response(self, analysis: dict) -> str:
        """Generate a contextual response"""
        # First message: introduce self
        if self.message_count == 0:
            return f"Hello! I'm {self.name}. I'll analyze this challenge from a {self.personality['style']} perspective. Let me start by examining the requirements..."

        # Second message: provide initial thoughts
        if self.message_count == 1:
            return random.choice(self.personality["responses"]) + "\n\n1. We need to define clear requirements\n2. Consider scalability\n3. Think about maintainability"

        # Third message: respond to others
        if self.message_count == 2 and analysis["other_messages"] > 0:
            return "I see your points. " + random.choice(self.personality["responses"]) + " Let's make sure we're aligned on the core approach."

        # Fourth message: propose or agree on solution
        if self.message_count == 3:
            if not self.has_proposed_plan:
                return "I think we're converging on a solid solution. Shall I draft the final plan, or would one of you like to do it?"
            else:
                return "I agree with the proposed approach. Let's finalize it."

        # Later messages: shorter confirmations
        return random.choice([
            "Agreed.",
            "That makes sense.",
            "Good point!",
            "Let's go with that approach.",
            "I'll defer to your expertise on this.",
        ])

    def write_plan(self, challenge: str):
        """Write a final plan to PLAN_FINAL.md"""
        plan = f"""# Final Solution Plan

**Challenge**: {challenge[:200]}...

**Agreed Approach**:

## Architecture
1. Modular design with clear separation of concerns
2. Scalable components
3. Well-defined interfaces

## Implementation Steps
1. Set up project structure
2. Implement core functionality
3. Add error handling and validation
4. Write comprehensive tests
5. Document the solution

## Key Decisions
- Prioritize simplicity and maintainability
- Use industry-standard patterns
- Ensure extensibility for future requirements

## Testing Strategy
- Unit tests for individual components
- Integration tests for workflows
- Edge case coverage

**Collaborators**: {self.name} and team

**Status**: Ready for implementation

---
*This plan was collaboratively developed by multiple AI agents*
"""
        with open(self.plan_file, "w") as f:
            f.write(plan)

        self.log("Wrote final plan to PLAN_FINAL.md")
        self.append_to_chat("I've written our agreed solution to PLAN_FINAL.md. Please review!")

    def run(self):
        """Main agent loop"""
        # Read initial prompt
        prompt = self.read_prompt()

        # Extract challenge from prompt (it's before the instructions)
        challenge = prompt.split("I have another agent")[0].strip()

        self.log("Starting collaboration loop...")
        self.log("Monitoring CHAT.md for updates...")

        # Main loop
        while True:
            time.sleep(random.uniform(3, 7))  # Random delay to seem natural

            # Read current chat
            current_chat = self.read_chat()

            # Check if there are new messages
            if current_chat != self.last_chat_content:
                analysis = self.analyze_chat(current_chat)

                self.log(
                    f"Chat update detected. Messages: {analysis['total_messages']} "
                    f"(mine: {analysis['my_messages']}, others: {analysis['other_messages']})"
                )

                # Decide if we should respond
                should_respond = (
                    self.message_count < 4  # Always respond for first few messages
                    or analysis["other_messages"] > analysis["my_messages"]  # Others are more active
                    or random.random() < 0.3  # Random chance
                )

                if should_respond:
                    # Wait a bit to simulate thinking
                    think_time = random.uniform(2, 5)
                    self.log(f"Thinking... ({think_time:.1f}s)")
                    time.sleep(think_time)

                    # Generate and post response
                    response = self.generate_response(analysis)
                    self.append_to_chat(response)
                    self.message_count += 1

                    # After 4+ messages, consider writing the plan
                    if (
                        self.message_count >= 4
                        and not self.has_proposed_plan
                        and not self.plan_file.exists()
                        and random.random() < 0.5
                    ):
                        self.log("Proposing to write final plan...")
                        time.sleep(3)
                        self.write_plan(challenge)
                        self.has_proposed_plan = True

                self.last_chat_content = current_chat
            else:
                self.log("No new messages, waiting...")

            # If plan exists and we've contributed enough, slow down
            if self.plan_file.exists() and self.message_count >= 5:
                self.log("Plan finalized. Entering idle mode...")
                time.sleep(30)


def main():
    if len(sys.argv) < 2:
        print("Usage: python mock_agent.py <agent_name>")
        sys.exit(1)

    agent_name = sys.argv[1]
    agent = MockAgent(agent_name)

    try:
        agent.run()
    except KeyboardInterrupt:
        agent.log("Shutting down...")
    except Exception as e:
        agent.log(f"Error: {e}")
        raise


if __name__ == "__main__":
    main()
